bundle/jsky.app.ot/src/main/scala/jsky/app/ot/viewer/SearchPanel.scala

```diff
@@ -0,0 +1,120 @@
+package jsky.app.ot.viewer
+
+import edu.gemini.pot.sp.{ISPNode, ISPObservation}
+import edu.gemini.spModel.data.ISPDataObject
+import edu.gemini.spModel.rich.pot.sp._
+
+import javax.swing.JOptionPane
+import javax.swing.event.DocumentEvent
+import javax.swing.event.DocumentListener
+
+import scala.swing._
+import scala.swing.GridBagPanel._
+import scala.swing.Swing.EmptyBorder
+import scala.swing.event.FocusLost
+
+import scalaz._
+import Scalaz._
+
+/** Defines a label and text box to use for searching for program nodes.
+  */
+class SearchPanel(statusFilter: StatusFilter, tree: SPTree) extends GridBagPanel {
+  import SearchPanel._
+
+  private var viewer = Option.empty[SPViewer]
+
+  def searchString: Option[String] =
+    Option(textBox.text).filterNot(_.isEmpty)
+
+  object nextAction extends Action("Next") {
+
+    def apply: Unit = {
+      val focus = for {
+        v     <- viewer
+        s     <- searchString
+        t     <- Option(v.getTree)
+        n     <- Option(t.getCurrentNode)
+        p     <- Option(n.getProgram)
+        (l, r) = p.toStream.span(_.getNodeKey =/= n.getNodeKey)
```

Comment:
I'm a bit confused. Isn't `p` an `ISPProgram`? I'm not sure what the `toStream` method is acting on or where it is coming from.

This looks good to me. As @tpolecat said, very nice, elegant solution for the search.